syntax = "proto3";
package ProjectJ;

import "Enum.proto";
import "Struct.proto";

message C_VERIFY_TOKEN
{
	int32 account_id = 1;
	string token = 2;
}

message S_VERIFY_TOKEN
{
	bool result = 1;
}

/* LOBBY PROTOCOL */

message C_LOBBY_CHAT
{
	int32 account_id = 1;
	string nickname = 2;
	string chat = 3;
}

message S_LOBBY_CHAT
{
	int32 account_id = 1;
	string nickname = 2;
	string chat = 3;
}

message C_LOBBY_REFRESH_ROOM
{
	int32 account_id = 1;
}

message S_LOBBY_REFRESH_ROOM
{
	repeated Room rooms = 1;
}

message C_LOBBY_CREATE_ROOM
{
	int32 account_id = 1;
	string title = 2;
}

message S_LOBBY_CREATE_ROOM
{
	bool result = 1;
	RoomInfo info = 2;
}

message C_LOBBY_ENTER_ROOM
{
	int32 account_id = 1;
	int32 room_id = 2;
}

message S_LOBBY_ENTER_ROOM
{
	bool result = 1;
	int32 room_id = 2;
	RoomInfo info = 3;
}

message C_ROOM_LEAVE
{
	int32 account_id = 1;
	int32 room_id = 2;
}

message S_ROOM_LEAVE
{
	bool result = 1;
}

message S_ROOM_OTHER_ENTER
{
	Player other = 1;
	RoomInfo info = 2;
}

message S_ROOM_OTHER_LEAVE
{
	Player other = 1;
	RoomInfo info = 2;
}

message C_ROOM_READY
{
	int32 account_id = 1;
	int32 room_id = 2;
}

message S_ROOM_READY
{
	RoomInfo info = 1;
}

message C_ROOM_CHAT
{
	int32 account_id = 1;
	string nickname = 2;
	int32 room_id = 3;
	string chat = 4;
}

message S_ROOM_CHAT
{
	int32 account_id = 1;
	string nickname = 2;
	int32 room_id = 3;
	string chat = 4;
}

message S_ROOM_STANDBY_MATCH
{
	uint32 count = 1;
}

message S_ROOM_START_MATCH
{
	bool start = 1;
}

/* MATCH PROTOCOL */

message C_MATCH_READY_TO_RECEIVE
{
	int32 account_id = 1;
}

message S_MATCH_ALL_LOADING_COMPLETE
{
	int32 your_player_index = 1;
	MatchInitInfo info = 2;
}

message S_MATCH_START
{
	uint64 end_tick = 1;
	uint64 current_tick = 2;
}

message C_MATCH_INFO
{
	message MyInfo
	{
		Vector position = 1;
		Rotator rotation = 2;
	}

	int32 player_index = 1;
	MyInfo info = 2;
}

message S_MATCH_INFO
{
	uint64 current_tick = 1;
	MatchInfo info = 2;
}

message S_MATCH_END
{
	int32 chaser_score = 1;
	int32 fugitivie_first_score = 2;
	int32 fugitivie_second_score = 3;
	int32 fugitivie_third_score = 4;
}

message C_MATCH_ITEM_PICKUP
{
	int32 player_index = 1;
	int32 item_index = 2;
}

message C_MATCH_ITEM_MOVE
{
	int32 player_index = 1;
	int32 from_index = 2;
	int32 to_index = 3;
	int32 item_index = 4;
	int32 target_top_left_index = 5;
	bool is_item_rotated = 6;
}

message C_MATCH_ITEM_DROP
{
	int32 player_index = 1;
	int32 item_index = 2;
	Vector drop_item_position = 3;
	Rotator drop_item_rotation = 4;
}

message S_MATCH_ITEM_SOMEONE_PICKUP
{
	int32 player_index = 1;
	int32 item_index = 2;
	int32 top_left_index = 3;
	bool is_item_rotated = 4;
}

message S_MATCH_ITEM_SOMEONE_MOVE
{
	int32 player_index = 1;
	int32 from_index = 2;
	int32 to_index = 3;
	int32 item_index = 4;
	int32 top_left_index = 5;
	bool is_item_rotated = 6;
}

message S_MATCH_ITEM_SOMEONE_DROP
{
	int32 player_index = 1;
	int32 item_index = 2;
	Vector drop_item_position = 3;
	Rotator drop_item_rotation = 4;
}



